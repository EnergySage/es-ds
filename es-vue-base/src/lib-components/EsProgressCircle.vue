<template>
    <div
        v-bind="$attrs"
        v-on="$listeners">
        <!-- Based on prior art from es-cdgm codebase -->
        <svg
            viewBox="0 0 100 100"
            :height="height"
            :width="height">
            <circle
                class="inner-circle"
                :cx="50"
                :cy="50"
                stroke-width="8"
                r="45%" />
            <!-- The first argument in `stoke-dasharray` is the amount of the circle we want colored
                The second is the gap before the next stroke starts
                Since we don't want another stroke to start, we use a number greater than the
                circumference of the circle (ex: 300). -->
            <template v-if="fillLength > 0">
                <circle
                    class="progress-circle"
                    :stroke-dasharray="`${fillLength} 300`"
                    :cx="50"
                    :cy="50"
                    r="45%"
                    stroke-width="8"
                    stroke-linecap="round"
                    transform="rotate(-90, 50, 50)" />
            </template>
            <template v-if="showPercentage">
                <text
                    x="50%"
                    y="52%"
                    class="font-weight-boldest text-gray-700 font-size-base"
                    dominant-baseline="middle"
                    alignment-baseline="middle"
                    text-anchor="middle">
                    {{ value }}%
                </text>
            </template>
        </svg>
    </div>
</template>

<script lang="js">

export default {
    name: 'EsProgressCircle',
    props: {
        height: {
            type: String,
            default: '50px',
        },
        value: {
            type: Number,
            required: true,
        },
        circle: {
            type: Boolean,
            default: false,
        },
        showPercentage: {
            type: Boolean,
            default: true,
        },
    },
    computed: {
        /**
         * @returns {number} the length of the stroke that displays the part of the circle
         * that is proportional to the value passed in.
         */
        fillLength() {
            return this.computedHeight * Math.PI * (this.value / 100) * (90 / this.computedHeight);
        },
        computedHeight() {
            const height = this.height.replace(/[^0-9]/g, '');
            return parseInt(height, 10);
        },
    },
};
</script>

<style lang="scss" scoped>
@use '~@energysage/es-bs-base/scss/bootstrap.import' as *;
// @import '~@energysage/es-bs-base/scss/includes';

.inner-circle {
    // position: relative;
    fill: transparent;
    stroke: $gray-200;
}

.progress-circle {
    // position: relative;
    fill: transparent;
    stroke: $primary;
}

</style>
