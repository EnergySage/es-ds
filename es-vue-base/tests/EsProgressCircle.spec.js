import { mount } from '@vue/test-utils';
import EsProgressCircle from '@/src/lib-components/EsProgressCircle.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsProgressCircle', () => {
    test('<EsProgressCircle />', async () => {
        const wrapper = mount(EsProgressCircle, {
            ...jestVue,
            propsData: {
                value: 55,
            },
        });

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('check for <EsProgressCircle />', async () => {
        const wrapper = mount(EsProgressCircle, {
            ...jestVue,
            propsData: {
                value: 55,
            },
        });

        expect(wrapper.find('circle').exists()).toBe(true);
    });

    // test default prop values
    test('<EsProgressCircle /> default prop values', async () => {
        const wrapper = mount(EsProgressCircle, {
            ...jestVue,
            propsData: {
                value: 55,
            },
        });
        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();

        expect(wrapper.props('showPercentage')).toBe(true);
    });

    test('<EsProgressCircle /> updates prop values', async () => {
        const wrapper = mount(EsProgressCircle, {
            ...jestVue,
            propsData: {
                showPercentage: false,
                value: 21,
            },
        });
        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();

        expect(wrapper.props('showPercentage')).toBe(false);
    });
});
