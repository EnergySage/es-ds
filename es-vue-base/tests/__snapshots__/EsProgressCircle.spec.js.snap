// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsProgressCircle <EsProgressCircle /> 1`] = `
"<div><svg viewBox="0 0 100 100" height="50px" width="50px">
    <circle cx="50" cy="50" stroke-width="8" r="45%" class="inner-circle"></circle>
    <circle stroke-dasharray="155.50883635269477 300" cx="50" cy="50" r="45%" stroke-width="8" stroke-linecap="round" transform="rotate(-90, 50, 50)" class="progress-circle"></circle> <text x="50%" y="52%" dominant-baseline="middle" alignment-baseline="middle" text-anchor="middle" class="font-weight-boldest text-gray-700 font-size-base">
      55%
    </text>
  </svg></div>"
`;

exports[`EsProgressCircle <EsProgressCircle /> default prop values 1`] = `
"<div><svg viewBox="0 0 100 100" height="50px" width="50px">
    <circle cx="50" cy="50" stroke-width="8" r="45%" class="inner-circle"></circle>
    <circle stroke-dasharray="155.50883635269477 300" cx="50" cy="50" r="45%" stroke-width="8" stroke-linecap="round" transform="rotate(-90, 50, 50)" class="progress-circle"></circle> <text x="50%" y="52%" dominant-baseline="middle" alignment-baseline="middle" text-anchor="middle" class="font-weight-boldest text-gray-700 font-size-base">
      55%
    </text>
  </svg></div>"
`;

exports[`EsProgressCircle <EsProgressCircle /> updates prop values 1`] = `
"<div><svg viewBox="0 0 100 100" height="50px" width="50px">
    <circle cx="50" cy="50" stroke-width="8" r="45%" class="inner-circle"></circle>
    <circle stroke-dasharray="59.37610115284709 300" cx="50" cy="50" r="45%" stroke-width="8" stroke-linecap="round" transform="rotate(-90, 50, 50)" class="progress-circle"></circle>
    <!---->
  </svg></div>"
`;
